### PR TITLE
Playing around: make unit types smarter

### PR DESCRIPTION
**Pull request by bstoeger**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

The unit types weren't the worst idea, as they make their base type explicit to the user. However, they were used very inconsistently, presumably because they are a pain to use.

This PR adds convenience operators so that the user needn't know the base type when adding  / subtracting.

The biggest gain will probably be comparison function, but for those I want to wait until we up the compiler version to C++20. Then one has to only implement a single `<=>` operator (a.k.a the space ship operator) instead of (`=`, `!=`, `<`, `>`, `<=`, `>=`). Then one should be able to use other convenience functions, such as `std::min`, `std::max` and `std:::clamp`.

Multiplication / division by scalars should obviously also be supported.

Note: related to #4237.

### CODE REVIEW

File: core/dive.cpp [hunk 1]
```diff
@@ -1148,12 +1148,12 @@ static void merge_one_sample(const struct sample &sample, struct divecomputer &d
...
 			/* Init a few values from prev sample to avoid useless info in XML */
 			surface.bearing.degrees = prev.bearing.degrees;
-			surface.ndl.seconds = prev.ndl.seconds;
+			surface.ndl = prev.ndl;
 			surface.time.seconds = last_time + 20;
 
 			append_sample(surface, &dc);
 
-			surface.time.seconds = sample.time.seconds - 20;
+			surface.time -= duration_t { .seconds = 20 };
```
Comment (mikeller) [inline core/dive.cpp]: This seems to be incorrect as it removes `sample.time` from the assignment.

File: core/load-git.cpp [hunk 2]
```diff
@@ -105,6 +105,11 @@ static pressure_t get_pressure(const char *line)
 	return p;
 }
 
+static o2pressure_t get_o2pressure(const char *line)
+{
+	return o2pressure_t { .mbar = static_cast<uint16_t>(lrint(1000 * ascii_strtod(line, NULL))) };
```
Comment (mikeller) [inline core/load-git.cpp]: Should `get_pressure()` use the same pattern for consistency?

File: core/dive.cpp (hunk 1 above)
Comment (mikeller) [inline core/dive.cpp]: I think this is wrong, as it removes `sample.time` from the assignment.

File: core/load-git.cpp [hunk 3]
```diff
@@ -105,6 +105,11 @@ static pressure_t get_pressure(const char *line)
 	return p;
 }
 
+static o2pressure_t get_o2pressure(const char *line)
```
Comment (mikeller) [inline core/load-git.cpp:104]: Should `get_pressure()` be using the same pattern as this?

File: core/sample.h [hunk 4]
```diff
@@ -21,7 +21,7 @@ struct sample                         // BASE TYPE BYTES  UNITS    RANGE
 	pressure_t pressure[MAX_SENSORS]; // int32_t  2x4    mbar   (0-2 Mbar)             cylinder pressures (main and CCR o2)
 	o2pressure_t setpoint;            // uint16_t   2    mbar   (0-65 bar)             O2 partial pressure (will be setpoint)
 	o2pressure_t o2sensor[MAX_O2_SENSORS];// uint16_t 6x2    mbar   (0-65 bar)             Up to 6 PO2 sensor values (rebreather)
-	bearing_t bearing = { -1 };       // int16_t    2  degrees  (-1 no val, 0-360 deg) compass bearing
+	bearing_t bearing = { .degrees = -1 };// int16_t    2  degrees  (-1 no val, 0-360 deg) compass bearing
```
Comment (mikeller) [inline core/sample.h:24]: This is an example for where the 'vertical alignment' whitespace pattern comes apart for me - if we'd wanted to stick to the pattern we'd need to add 20+ lines with whitespace only changes to the pull request.

**Review by mikeller — COMMENTED** [review]
See my comments for some potential bugs and possible improvements.

All in all I am not the biggest fan of the structs in `unit.h` - to me they are providing little benefit over being explicit with units when naming variables (like `time_s`), and have the potential to lead to errors when they are used inconsistently like before this pull request, like when values in mbar are implicitly initialised with a value in bar.

So this is definitely going into the right direction.

**Review by mikeller — APPROVED** [review]

File: core/dive.cpp (hunk 1 above)
Comment (bstoeger) [reply core/dive.cpp]: Yes, I misread sample as surface.

**Comment by bstoeger** [comment]
> See my comments for some potential bugs and possible improvements.
> 
> All in all I am not the biggest fan of the structs in `unit.h` - to me they are providing little benefit over being explicit with units when naming variables (like `time_s`), and have the potential to lead to errors when they are used inconsistently like before this pull request, like when values in mbar are implicitly initialised with a value in bar.
> 
> So this is definitely going into the right direction.

I like the idea - forcing a base type and making it explicit when setting/reading seems reasonable.

However, the current implementation and usage is horrid. It completely defeats the purpose of a unit type if one has to state the base type for comparing and doing arithmetic.

**Comment by glance-** [comment]
I really liked the declared types, and it made it way easier to ensure you actually passed the right types to the right functions. Sure, it made most sense in C code, where you'd might have had int get_usable_volume(int he, int o2, int cylinder_size, int pressure) and you easily could have mixed up the order. When doing it with volume_t get_usable_volume(struct gasmix mix, volume_t cylinder_size, presure_t pressure) you'd get a compiler error if you mixed them up.
There have been quite a few such bugs in this projects history.

Also, the fact that it's easy to see in which unit this depth, pressure and volume are have for me been a great helper so one doesn't mix them up. 

Now when we're moving to c++, it also makes sense to implement sane operations like these with types and between types where they make sense.

**Review by mikeller — APPROVED** [review]
Looks good to me.

**Comment by bstoeger** [comment]
Opinions on the user-defined literals? @dirkhh? @atdotde?

**Comment by dirkhh** [comment]
I'm not sure what you mean by user defined literals.
Things like 0_m and 0_bar?

**Comment by bstoeger** [comment]
Yes, precisely: https://en.cppreference.com/w/cpp/language/user_literal

**Comment by dirkhh** [comment]
Let's add this to the long list of "this is weird" and go on with it. I can never see me wrapping my mind around this, but I understand that this makes sense to people, so go ahead.

**Comment by bstoeger** [comment]
I think it's a neat convenience feature and makes the code more readable. But it isn't important to me.

**Comment by dirkhh** [comment]
Maybe my point is hidden behind my... writing patterns.
I agree that this is useful. I understand that for people who are not me this creates code that is easier to read.
So I was encouraging you to include this.

**Comment by bstoeger** [comment]
> Maybe my point is hidden behind my... writing patterns.

No no, I was just giving my personal opinion pretty much independently of yours.

**Comment by mikeller** [comment]
> I think it's a neat convenience feature and makes the code more readable. But it isn't important to me.

Agreed re readability.
I think the one thing that we need to keep in mind with operator overloads like these ones is that this can be confusing and hard to follow for casual contributors:
If they come across these constructs in the file they want to modify, it will not be obvious how it is done, and finding the location where the overload is defined will be hard for them, without knowledge of the code base.
Not sure if there is much we can do to help with this.

**Comment by bstoeger** [comment]
Yes, at the very least there should be comments in units.h... Perhaps there also should be a chapter on the unit-types in the coding-style document.

**Comment by mikeller** [comment]
> Yes, at the very least there should be comments in units.h... Perhaps there also should be a chapter on the unit-types in the coding-style document.

Good idea to put it into `CODINGSTYLE.md`. Having an explanation in `units.h` is good as well, but at least for new users it will be hard to figure out that this is where to look for this.

File: core/dive.cpp [hunk 5]
```diff
@@ -2399,7 +2401,7 @@ bool dive::cache_is_valid() const
 pressure_t dive::get_surface_pressure() const
 {
 	return surface_pressure.mbar > 0 ? surface_pressure
-					 : pressure_t { SURFACE_PRESSURE };
+					 : pressure_t { .mbar = SURFACE_PRESSURE };
```
Comment (mikeller) [inline core/dive.cpp]: While we are playing with unit types, should we replace / augment `SURFACE_PRESSURE` with something that allows us to flex our new unit system when using it?

**Comment by bstoeger** [comment]
Let's commit so that I can rebase the other stuff without things diverging too much.

I'm still on the fence concerning the comparison operators. Switch to C++20 and implement a single `<=>` or do all the annoying `==`, `!=`, `<`, `>`, `<=`, `>=`?

**Comment by dirkhh** [comment]
Just make sure you don't needless break older builds (like the ones for macOS 10.15). I don't think that compiler that I need to use for that supports C++20
